File: di_java/src/test/java/com/kmmania/hr/ServiceImplHRTest.java
package com.kmmania.hr;

import com.kmmania.api.DAO;
import com.kmmania.api.Service;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

class ServiceImplHRTest {
    @InjectMocks
    private Service serviceMock;

    @Mock
    private DAO daoMock;

    @BeforeEach
    void setUp() {
        daoMock = mock(DAOImplHR.class);
        serviceMock = mock(ServiceImplHR.class);
    }

    @Test
    @DisplayName("compute() method should compute data from DAO")
    void computeShouldComputeDataFromDAO() {
        // Given
        when(daoMock.getData()).thenReturn(185.0);
        when(serviceMock.compute()).thenReturn(138.75);

        // When
        double resultDAO = daoMock.getData();
        double resultService = serviceMock.compute();

        // Then
        verify(daoMock, times(1)).getData();
        assertThat(resultDAO).isEqualTo(185.0); // Expected HR max value provided by DAO
        verify(serviceMock, times(1)).compute();
        assertThat(resultService).isEqualTo(138.75); // Expected result is 75% of 185
    }
}

File: di_kotlin/src/test/kotlin/hr/ServiceImplHRTest.kt
package hr

import com.kmmania.api.DAO
import com.kmmania.api.Service
import com.kmmania.hr.DAOImplHR
import com.kmmania.hr.ServiceImplHR
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.*

class ServiceImplHRTest {
    @InjectMocks
    private lateinit var serviceMock: Service

    @Mock
    private lateinit var daoMock: DAO

    @BeforeEach
    fun setUp() {
        daoMock = mock(DAOImplHR::class.java)
        serviceMock = mock(ServiceImplHR::class.java)
    }

    @Test
    fun `compute() method should compute data from DAO`() {
        // Given
        `when`(daoMock.getData()).thenReturn(185.0)
        `when`(serviceMock.compute()).thenReturn(138.75)

        // When
        val resultDao = daoMock.getData()
        val resultHr = serviceMock.compute()

        // Then
        verify(daoMock, times(1)).getData()
        assertThat(resultDao).isEqualTo(185.0)
        verify(serviceMock, times(1)).compute()
        assertThat(resultHr).isEqualTo(138.75)
    }
}